### ISSUE TITLE
Accessible content in balloons-and-static-electricity

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/balloons-and-static-electricity#103]
Balloons and Static Electricity is being used as a testing bed for accessible content, specifically keyboard navigation and auditory description through screen readers.  See issue #100.  

Creating this issue to document progress and challenges faced in the implementation.

**Comment by jessegreenberg** [comment]
See #102 for a description for challenges related to RadioButtonGroups.

**Comment by jessegreenberg** [comment]
Initial keyboard navigation for this sim is complete.  The following is the rendered accessibility tree:

```html
<div class="accessibility" style="position: absolute; left: 0px; top: 0px; width: 0px; height: 0px; clip: rect(0px 0px 0px 0px);">
	<div class="ScreenView">
		<header role="banner" aria-labelledby="scene-label-14-35" aria-describedby="scene-descripion-14-35">
			<h1 id="scene-label-14-35">‪The scene for Balloons and Static Electricity‬</h1>
			<div id="scene-descripion-14-35">
				<p>‪A balloon floats evenly between a large woolly sweater and a removable wall. The sweater is to the left and the wall is to the right.  Each object shows negative and positive charges. Charges are shown in pairs. Currently, the sweater and the wall have many pairs of charges, the balloon, just a few. The control panel has controls to add a second balloon, adjust charge views and to reset the entire experiment. Select Tab for next item. Select K or question mark for keyboard help. Select H for a heading outline.‬</p>
			</div>
		</header>
		<div class="ScreenViewContainer" role="main">
			<section id="play-area-14-35-37" aria-labelledby="heading-node-36">
				<h2 id="heading-node-14-35-37-36">‪Play Area‬</h2>
				<div id="sweater-14-35-37-38" aria-labelledby="sweater-label-14-35-37-38" aria-describedby="sweater-description-14-35-37-38">
					<h3 id="sweater-label-14-35-37-38">‪Sweater‬</h3>
					<p id="sweater-description-14-35-37-38">‪The sweater has a neutral charge, no more positive charges than negative ones.‬</p>
				</div>
				<div id="wall-14-35-37-156" aria-labelledby="[object HTMLHeadingElement]">
					<h3 id="wall-label-14-35-37-156">‪Wall‬</h3>
					<p id="wall-description-14-35-37-156">‪The wall has a neutral charge, no more positive charges than negative ones.‬</p>
				</div>
				<button role="slider" title="Balloon" id="balloon-14-35-37-270-342" draggable="true" class="Balloon" aria-labelledby="balloon-label-14-35-37-270-342" aria-describedby="balloon-description-14-35-37-270-342">
					<h3 id="balloon-label-14-35-37-270-342">‪Balloon one‬</h3>
					<div id="balloon-description-14-35-37-270-342">
						<p>‪The balloon has a neutral charge, no more negative charges than positive ones. Currently, positioned at equal distance between sweater and wall. The sweater is to the left and the wall is to the right. The top of the balloon is level with the chest area of the sweater. Select any Arrow key to grab the balloon and start dragging. Select Tab for next item. Select K or question mark for keyboard help.‬</p>
					</div>
				</button>
				<button role="slider" title="Balloon" id="balloon-14-35-37-270-271" draggable="true" class="Balloon" hidden aria-labelledby="balloon-label-14-35-37-270-271" aria-describedby="balloon-description-14-35-37-270-271">
					<h3 id="balloon-label-14-35-37-270-271">‪Balloon two‬</h3>
					<div id="balloon-description-14-35-37-270-271">
						<p>‪The balloon has a neutral charge, no more negative charges than positive ones. Currently, positioned at equal distance between sweater and wall. The sweater is to the left and the wall is to the right. The top of the balloon is level with the chest area of the sweater. Select any Arrow key to grab the balloon and start dragging. Select Tab for next item. Select K or question mark for keyboard help.‬</p>\
					</div>
				</button>
			</section>
			<section id="control-panel-14-35-413" aria-labelledby="heading-node-414">
				<h2 id="heading-node-14-35-413-414">‪Control Panel‬</h2>
				<input type="button" value="‪Remove Wall‬" aria-describedby="reset-all-description-14-35-413-485-427"></input>
				<fieldset id="radio-button-group-14-35-413-488-480-470" role="radiogroup" aria-describedby="radio-group-description-14-35-413-488-480-470">
					<legend id="legend-14-35-413-488-480-470-490">‪Balloon Controls‬</legend>
					<input id="radio-button-14-35-413-488-480-470-464" type="radio" role="radio" name="16" aria-label="‪Remove green balloon‬" aria-checked="true">
					<input id="radio-button-14-35-413-488-480-470-467" type="radio" role="radio" name="16" aria-label="‪Add green balloon‬" aria-checked="false">
					<p id="radio-group-description-14-35-413-488-480-470">‪Toggle to conduct experiments with two balloons or just one.‬</p>
				</fieldset>]
				<input type="button" value="‪Reset Balloons‬" aria-describedby="reset-all-description-14-35-413-488-480-477"></input>
				<fieldset id="radio-button-group-14-35-413-488-486-457" role="radiogroup" aria-describedby="radio-group-description-14-35-413-488-486-457">
					<legend id="legend-14-35-413-488-486-457-489">‪Charge Settings‬</legend>
					<input id="radio-button-14-35-413-488-486-457-439" type="radio" role="radio" name="11" aria-label="‪Show all charges‬" aria-checked="true">
					<input id="radio-button-14-35-413-488-486-457-447" type="radio" role="radio" name="11" aria-label="‪Show no charges‬" aria-checked="false">
					<input id="radio-button-14-35-413-488-486-457-455" type="radio" role="radio" name="11" aria-label="‪Show charge differences‬" aria-checked="false">
					<p id="radio-group-description-14-35-413-488-486-457">‪Select how you want to see or hear the charges.‬</p>
				</fieldset>
				<input type="reset" value="‪Reset All‬" aria-describedby="reset-all-description"></input>
			</section>
		</div>
	</div>
</div>
```

**Comment by jessegreenberg** [comment]
List of issues found while working today:
1) NVDA seems to cache the sim page in a way that prevents it from doing a full read through of the DOM after the page is refreshed.  I am not sure if this is an issue for other screen readers.

2) There is a delay as the page loads and NVDA doesn't wait for the page to load before starting to read.  This means that it reads through the entire URL because the title of the sim hasn't loaded yet.  Not sure if this is an issue for other AT.

3)  I found a bug in my html that was linking them all together so that the user could not navigate to different radio button groups.  This has been fixed.  See #102.

4) Giving the balloon the 'slider' role allows us to use arrow keys in documents mode.  Actually, the role forces the AT to switch to forms mode, but it will switch back when the user presses 'tab'.  Arrow keys and hot keys are working just fine!  There may be a better role than 'slider' for the balloon, but that one works for now.

5) Arrow keys do not seem to work for NVDA + Chrome.  Shift + arrow keys work for chrome, but arrow keys alone do not.  For all other platforms that I tested, the arrow keys are working.

6) Behavior seems much more predictable than our first iterations, and the user should be able to find headings and paragraphs with AT.

**Comment by jessegreenberg** [comment]
Notice two issues in above DOM:

Balloons and wall are out of order, and Wall 'aria-labelledby' id is invalid.  Will fix and then deploy a dev version.

**Comment by terracoda** [comment]
@jessegreenberg 
Re P-DOM and Balloon sliders. The title attribute for both balloons is the same. I don’t think you need the title attribute, but if you want to keep it, it should be “Yellow Balloon” and “Green Balloon”. This would provide extra information that is relevant to both visual and non-visual users and would connect "Balloon One" with “Yellow Balloon”.
Also, now that you are using a button slider for the balloon, I’m not sure we need the heading h3. I don’t think it is valid HTML to put all that stuff inside a button?

**Comment by jessegreenberg** [comment]
Good catch @terracoda.  I took the title from an example I found of ```role="slider"```, but I think it is unnecessary now.

I thought the ```H3``` for the balloon was still valuable because the user could quickly find the element and its description with the AT.  You are right, I don't think the button should have children.  

Now that we are using the slider role, it may be interactive enough to contain all balloon elements in a div.

**Comment by jessegreenberg** [comment]
Other issues and feature requests noted by @terracoda:

- [ ] For an initial round of student testing, dynamic text descriptions should be removed.  'Wizard of Oz' testing method will be used.

- [ ] ChromeVox did not read screen description, only the H1 element for the screen view.

- [ ] Arrow Key movement of the balloon needs to be faster.

- [ ] Vertical rubbing of the balloon does not seem to pick up charges.  Only horizontal translation works.

- [ ] [JN] 'Jump Near' hot key doesn't take the balloon quite to the right 'sweet' spot.  Balloon should go back to sweater when released.

- [ ] 'Ctrl + Enter' does not release the balloon on a mac.

- [ ] Wall Button accessible description needs to change when pressed to match screen description.

- [ ] Balloon focus highlight should be just around the balloon, not the balloon and its string.

- [ ] The focus highlight of the balloon needs to change. Pink for default, black when grabbed.

- [ ] 'Play Area' heading is ordered after 'Wall' in some browsers.

- [ ]  Inputs should be self closing.

**Comment by jessegreenberg** [comment]
698aeb13534cee35ec10b7ea9ca55dfe71c8b446 resolves https://github.com/phetsims/balloons-and-static-electricity/issues/103#issuecomment-164596237.

**Comment by jessegreenberg** [comment]
> Vertical rubbing of the balloon does not seem to pick up charges. Only horizontal translation works.

By design, only the left side of the balloon picks up charges.

```js
//is balloon over minus charge on sweater?
    findIntersection: function( balloon ) {
      var self = this;
      //active area of balloon rectangle at the left of image
```

**Comment by jessegreenberg** [comment]
> Sweater heading and description are empty

Found references of innerText, not supported by some browsers.  Replaced with textContent.

**Comment by jessegreenberg** [comment]
> Add invisible keyboard help dialog

Dialogs will be tricky.  See #105.

**Comment by jessegreenberg** [comment]
Invisible dialog for keyboard interactions added, with some caveats.  See #105.

**Comment by jessegreenberg** [comment]
New dev version published.

**Comment by jessegreenberg** [comment]
This issue is a bit stale and we are tracking progress of accessible content in individual issues at this point.  Closing.
